woocommerce/woocommerce
# Expose Product Collection location in attributes

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





In https://github.com/woocommerce/woocommerce/pull/43997 we added a location recognition for Product Collection block. This PR builds on top of that:
- move the `useGetLocation` hook from Product Template utils to base-hooks
- move usage of the hook from Product Template to Product Collection
- assign the location data to attributes
- expose the location through context

Note: this PFR doesn't change the logic of `useGetLocation` which was already tested in https://github.com/woocommerce/woocommerce/pull/43997 so no need to re-review the hook itself.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Steps to test:

1. Try adding Product Collection in multiple places and verify the proper attribute is set on Product Collection block
2. To verify that in Editor, switch from Visual Editor to Code Editor (three dots menu in top right corner -> Code Editor)
3. Look for "location" property like this:

<img width="797" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/d80a8082-d2a4-44c8-9f79-3a1de0db230f">

5. Here are some example scenarios, but try to break it as well by trying something else:

Scenario | Output
-- | --
Products by Category/Tag/Attribute template | `{"type":"archive","sourceData":{"taxonomy":"product_cat","termId":null}}`
Category: SPECIFIC_CATEGORY template | `{"type":"archive","sourceData":{"taxonomy":"product_cat","termId":SPECIFIC_CATEGORY_ID}}`
Product Collection in Single Product block in Category: SPECIFIC_CATEGORY template | `{"type":"product","sourceData":{"productId": SPECIFIC_PRODUCT_ID}}`
Product Collection in Single Product: SPECIFIC_PRODUCT template | `{"type":"product","sourceData":{"productId": PRODUCT_ID}}`
In the above scenario, add Single Product block and Product Collection inside | `{"type":"product","sourceData":{"productId":PRODUCT_ID_FROM_SINGLE_PRODUCT_BLOCK}}`
Product Collection in Mini Cart Contents |  `{"type":"cart","sourceData":{}}`
Page/Post | `{"type":"generic","sourceData":{}}`

6. Location type matches the closest context and ID is provided if available.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
